File: flutter_application_1_backend/src/models/Answer.js
const mongoose = require("mongoose");

const AnswerSchema = new mongoose.Schema(
  {
    body: { type: String, required: true },
    author: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
    post: { type: mongoose.Schema.Types.ObjectId, ref: "Post" },
    votes: [
      {
        user: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
        vote: { type: Number, enum: [1, -1] }
      }
    ],
    isAccepted: { type: Boolean, default: false }
  },
  { timestamps: true }
);

AnswerSchema.virtual('voteCount').get(function() {
  return this.votes.reduce((sum, v) => sum + v.vote, 0);
});

AnswerSchema.virtual('upvotes').get(function() {
  return this.votes.filter(v => v.vote === 1).length;
});

AnswerSchema.virtual('downvotes').get(function() {
  return this.votes.filter(v => v.vote === -1).length;
});

AnswerSchema.set('toJSON', { virtuals: true });
AnswerSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model("Answer", AnswerSchema);

File: flutter_application_1/lib/models/answer.dart
class Answer {
  final String id;
  final String body;
  final dynamic author;
  final String post;
  final Votes votes;
  final bool isAccepted;
  final DateTime createdAt;
  final DateTime updatedAt;

  Answer({
    required this.id,
    required this.body,
    required this.author,
    required this.post,
    required this.votes,
    required this.isAccepted,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Answer.fromJson(Map<String, dynamic> json) {
    return Answer(
      id: json['_id'] ?? '',
      body: json['body'] ?? '',
      author: json['author'],
      post: json['post'] ?? '',
      votes: Votes.fromJson(json['votes'] ?? {}),
      isAccepted: json['isAccepted'] ?? false,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'body': body,
      'post': post,
    };
  }

  String get authorName {
    if (author is Map) {
      return author['username'] ?? 'Unknown';
    }
    return 'Unknown';
  }

  String? get authorId {
    if (author is String) {
      return author;
    } else if (author is Map && author['_id'] != null) {
      return author['_id'];
    }
    return null;
  }
}

class Votes {
  final int upvotes;
  final int downvotes;

  Votes({
    required this.upvotes,
    required this.downvotes,
  });

  factory Votes.fromJson(Map<String, dynamic> json) {
    return Votes(
      upvotes: json['upvotes'] ?? 0,
      downvotes: json['downvotes'] ?? 0,
    );
  }

  int get total => upvotes - downvotes;
}
